### PR TITLE
Remove Cuda

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
- Nobody uses the cuda implementation
- We have to maintain cuda implementation
- The interfaces around cuda are making changes we want difficult

#### Summary of Changes
- Remove cuda from the code

Fixes #

### CODE REVIEW

**Comment by LegNeato** [comment]
Any interest in possibly switching to https://github.com/Rust-GPU/rust-gpu/ instead?

**Comment by apfitzge** [comment]
> Any interest in possibly switching to [Rust-GPU/rust-gpu](https://github.com/Rust-GPU/rust-gpu/) instead?

No, that doesn't solve anything here.

**Comment by shirotech** [comment]
@apfitzge so the `--cuda` flag is broken and doesn't work?

Question, nobody is using it because it doesn't work or it's not needed because the performance gain isn't much?

**Comment by apfitzge** [comment]
> @apfitzge so the `--cuda` flag is broken and doesn't work?
> 
> Question, nobody is using it because it doesn't work or it's not needed because the performance gain isn't much?

It's my understanding that it currently works, but the benefit of running it is less than a node with GPU would cost you, so nobody does.
The way in which we are storing `PacketBatch` is blocking other changes we want to make (non-inlined packets). If we were to make those changes all the CUDA stuff would need to be fixed to work with that...which seems like wasted effort if nobody uses it.

**Comment by steviez** [comment]
This seems like something that should be called out in the CHANGELOG:
https://github.com/anza-xyz/agave/blob/master/CHANGELOG.md

File: perf/src/recycled_vec.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_pinned_vec() {
```
Comment (vadorovsky) [inline perf/src/recycled_vec.rs]:
```suggestion
    fn test_recycled_vec() {
```

**Review by vadorovsky — COMMENTED** [review]
I'm yet to test it properly, but so far looks good, only one nit.

**Review by vadorovsky — CHANGES_REQUESTED** [review]
There are still some leftovers from CUDA in your branch:

* https://github.com/apfitzge/agave/blob/kill_cuda/ci/setup-new-buildkite-agent/setup-cuda.sh I think this file should go away
  * Then remove the call of this script from `setup_new_machine.sh` and `setup-partner-node.sh` as well.
* `ci/README` still mentions CUDA
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L39-L45
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L68-L72
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L110-L113
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L137-L148
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L162
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L168
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/README.md?plain=1#L182-L184
* I think we don't need this script as well? https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/net/scripts/enable-nvidia-persistence-mode.sh
* `ci/test-stable.sh` calls the script above and sets CUDA-related variables:
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/ci/test-stable.sh#L91-L103
* This comment needs to be updated https://github.com/apfitzge/agave/blob/kill_cuda/sdk/clock/src/lib.rs#L140
* `net/scripts/gce-provider.sh`
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/net/scripts/gce-provider.sh#L175-L187
* `net/scripts/ec2-provider.sh`
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/net/scripts/ec2-provider.sh#L183-L205
* `net/README.md`
  * https://github.com/apfitzge/agave/blob/8cd2531fc1b9575272ef9df26a3b096fac2749c0/net/README.md?plain=1#L206-L216

**Comment by apfitzge** [comment]
@vadorovsky Given your feedback, I'm going to talk with @yihau about removing all the gpu and cuda from scripts and CI before this PR lands. I'm not familiar with those so gonna work with him to remove all those and after it is removed from CI/scripts we can remove the implementation/support with this PR.

**Comment by apfitzge** [comment]
Rebased on @yihau's CI changes (🙏 ). 

Remaining cuda references:
```
$ rg "cuda"
CHANGELOG.md
31:  * Remove support for `--cuda` from `agave-validator`

net/scripts/gce-provider.sh
182:    # imageName="ubuntu-2004-focal-v20201211-with-cuda-1
```

File: perf/src/recycled_vec.rs (hunk 1 above)
Comment (vadorovsky) [reply perf/src/recycled_vec.rs]: @apfitzge This is not addressed yet :slightly_smiling_face:

File: perf/src/recycled_vec.rs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    }
+    fn set_recycler(&mut self, recycler: Weak<RecyclerX<Self>>) {
+        self.recycler = recycler;
+    }
+}
+
+impl<T: Clone + Default + Sized> From<RecycledVec<T>> for Vec<T> {
+    fn from(mut pinned_vec: RecycledVec<T>) -> Self {
+        pinned_vec.recycler = Weak::default();
+        std::mem::take(&mut pinned_vec.x)
```
Comment (vadorovsky) [inline perf/src/recycled_vec.rs]:
```suggestion
    fn from(mut recycled_vec: RecycledVec<T>) -> Self {
        recycled_vec.recycler = Weak::default();
        std::mem::take(&mut recycled_vec.x)
```

File: perf/src/recycled_vec.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    pub fn resize(&mut self, size: usize, elem: T) {
+        self.x.resize(size, elem);
+    }
+
+    pub fn append(&mut self, other: &mut Vec<T>) {
+        self.x.append(other);
+    }
+
+    pub fn append_pinned(&mut self, other: &mut Self) {
```
Comment (vadorovsky) [inline perf/src/recycled_vec.rs]:
```suggestion
    pub fn append_recycled(&mut self, other: &mut Self) {
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]
Last nits, otherwise looks great. Thanks!

File: perf/src/recycled_vec.rs (hunk 1 above)
Comment (apfitzge) [reply perf/src/recycled_vec.rs]: ac660edf75f449b6594fa1d502be6477e41bbbb6

File: perf/src/recycled_vec.rs (hunk 2 above)
Comment (apfitzge) [reply perf/src/recycled_vec.rs]: 7572911940b809571c8fc2c0141c0c46761a7b7c

File: perf/src/recycled_vec.rs (hunk 3 above)
Comment (apfitzge) [reply perf/src/recycled_vec.rs]: no consumers of this function - removed instead of renaming: 01e71cbf91372ee4af96dda330349d7df2983460

**Review by vadorovsky — DISMISSED** [review]
Looks good, thanks!

The question is when do we want to merge it. I'm still working on the `Packet<Bytes>` change (it's taking longer than I expected, sorry 😔) - should we wait until I'm done and therefore 100% sure that the zero-copy approach actually?

Ping @alessandrod

**Comment by steviez** [comment]
Started reviewing but didn't finish. But looks like there is a conflict that will require resolution anyways

> The question is when do we want to merge it. I'm still working on the `Packet<Bytes>` change (it's taking longer than I expected, sorry 😔) - should we wait until I'm done and therefore 100% sure that the zero-copy approach actually?

I think it would make sense to push this one first. Given that this change is largely removing code, it should mean less stuff you have to account for with your change Michal. And, I think we feel pretty confident with the Bytes approach.

I don't know exactly how the quinn/TPU integration looks, but at least for the TVU path, the `Bytes` approach should save us an allocation per shred (currently, each shred allocates owned memory). With current MNB load, that is approximately ~3k allocation per second, each in excess of ~1 kB. The number of shreds should go up as we increase CU limits

File: perf/src/recycled_vec.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    serde::{Deserialize, Serialize},
+    std::{
+        ops::{Index, IndexMut},
+        slice::{Iter, IterMut, SliceIndex},
+        sync::Weak,
+    },
+};
+
+// A vector wrapper where the underlying memory can be
+// recycled.
```
Comment (steviez) [inline perf/src/recycled_vec.rs]: nit: This could go on one line

File: entry/src/entry.rs [hunk 5]
```diff
@@ -346,43 +309,8 @@ impl EntryVerificationState {
         self.poh_duration_us
     }
 
-    pub fn finish_verify(&mut self, thread_pool: &ThreadPool) -> bool {
+    pub fn finish_verify(&mut self) -> bool {
```
Comment (steviez) [inline entry/src/entry.rs:312]: Thoughts on ripping out the start and finish function / naming ? With CPU, we finish it immediately so the name could really be `verify_transactions` instead of `start_verify_transactions`, and this function could go away. I could see an argument for deferring that to a subsequent PR too, lemme know what you think

File: perf/src/packet.rs [hunk 6]
```diff
@@ -21,28 +21,22 @@ pub const NUM_RCVMMSGS: usize = 64;
...
         Self { packets }
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
-        let packets = PinnedVec::with_capacity(capacity);
+        let packets = RecycledVec::with_capacity(capacity);
         Self { packets }
     }
 
-    pub fn new_pinned_with_capacity(capacity: usize) -> Self {
```
Comment (steviez) [inline perf/src/packet.rs:40]: Unless you were planning on doing it, consolidating the various constructors will be a nice follow-on PR here. Not sure if leaving those out was intentional or not, but think it makes sense to do outside of this PR

File: perf/src/recycled_vec.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct RecycledVec<T: Default + Clone + Sized> {
+    x: Vec<T>,
+    #[serde(skip)]
+    recycler: Weak<RecyclerX<RecycledVec<T>>>,
+}
+
+impl<T: Default + Clone + Sized> Reset for RecycledVec<T> {
+    fn reset(&mut self) {
+        self.resize(0, T::default());
```
Comment (steviez) [inline perf/src/recycled_vec.rs:24]: `PinnedVec` had this line too, but `Vec::clear()` is probably more appropriate here. Avoid the `T::default()` + clear does less stuff than resize (which will call truncate)

File: perf/src/sigverify.rs [hunk 8]
```diff
@@ -552,78 +509,10 @@ pub fn mark_disabled(batches: &mut [PacketBatch], r: &[Vec<u8>]) {
...
-            USE_NON_DEFAULT_STREAM,
-        );
-        if res != 0 {
-            trace!("RETURN!!!: {}", res);
-        }
-    }
-    trace!("done verify");
-    copy_return_values(sig_lens, &out, &mut rvs);
-    mark_disabled(batches, &rvs);
+    ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
```
Comment (steviez) [inline perf/src/sigverify.rs:500]: nit: Maybe separate from this PR, we can consolidate our use of `ed25519_verify_cpu` and `ed25519_verify` and deprecate one of them

File: perf/src/sigverify.rs [hunk 9]
```diff
@@ -687,8 +573,9 @@ mod tests {
                     .collect()
             })
             .collect();
-        let out =
-            PinnedVec::<u8>::from_vec(out.into_iter().flatten().flatten().map(u8::from).collect());
+        let out = RecycledVec::<u8>::from_vec(
+            out.into_iter().flatten().flatten().map(u8::from).collect(),
+        );
```
Comment (steviez) [inline perf/src/sigverify.rs]: The function exercised in this test, `copy_return_values()`, was only used in GPU path AFAIK. So, rip that function + this test out too ?

**Comment by behzadnouri** [comment]
> * Nobody uses the cuda implementation

That is the case "today".
Sigverify is a major bottleneck of the pipline, more so once there are more transactions, shreds, gossip, etc packets to sigverify.
What if the load increases so much that our hands are forced to use gpu for that?

> * The interfaces around cuda are making changes we want difficult

What changes specifically?
are those changes addressing bottlenecks more significant than sigverify?

**Comment by apfitzge** [comment]
> > * Nobody uses the cuda implementation
> 
> That is the case "today". Sigverify is a major bottleneck of the pipline, more so once there are more transactions, shreds, gossip, etc packets to sigverify. What if the load increases so much that our hands are forced to use gpu for that?
> 
> > * The interfaces around cuda are making changes we want difficult
> 
> What changes specifically? are those changes addressing bottlenecks more significant than sigverify?

Our hands will never be "forced" to use GPU because there are already better solutions than using a gpu for this.
Changes to make Packet not store the bytes inline, because right now everytime we move a `Packet` we're copying 1232 bytes, we want to not copy them.
If the bytes are not inline the current cuda code is broken.
IS the benefit of no-copy more significant than sigverify? No obviously not! and that's not the alternative, the alternative is just maintaining this code that no one uses.

If we have to re-add support for GPUs (unlikely) we can fix it then. There are other completely inefficient things the gpu impl is doing as well. it's not worth the cost of maintaining a feature no one uses. The code isn't going into the ether never to be seen again...we can always re-use parts of the current implementation iff we need to.

**Comment by behzadnouri** [comment]
> Our hands will never be "forced" to use GPU because there are already better solutions than using a gpu for this. 

What are the better solutions?

> Changes to make Packet not store the bytes inline, because right now everytime we move a `Packet` we're copying 1232 bytes, we want to not copy them. 

That would also break the recycler, and you may end up doing even more allocations or memcopies.
Shouldn't first confirm that that is a good idea before committing to it?!

> the alternative is just maintaining this code that no one uses.
> it's not worth the cost of maintaining a feature no one uses.

I am not sure how much time anyone has spend on maintaining cuda code in the past couple of years.

**Comment by apfitzge** [comment]
> What are the better solutions?

fpga or smart nic

> That would also break the recycler, and you may end up doing even more allocations or memcopies. Shouldn't first confirm that that is a good idea before committing to it?!

recycler really only made sense in the context of pinned memory. jemalloc already keeps caches of memory that it will re-use for the packets.

@alessandrod can probably list the many benefits in networking code.
In SV it will enable us to eventually move prioritization earlier, which is not possible when we have to move batches of packets around...which we're currently forced to do for performance if because packet data is inline. If the packets are just ptr + meta, then we can very cheaply move them around individually. 
In BS, it stops us from having to manage our own memory and frees up a ton of capacity because our already scheduled packets do not need to take up room in the scheduler buffer. 

> I am not sure how much time anyone has spend on maintaining cuda code in the past couple of years.

Nobody uses an umbrella until it rains. It's getting in the way now, and would require a large rewrite to make work - all of the indexing does not work if the packet memory is not inline. We're trying to make progress in making the chain better. Taking the time to do this properly slows that down for very little, if any, benefit.


edit: I apologize if my responses seem short or rude. This was all discussed on slack previously in a channel you are in.

File: entry/src/entry.rs (hunk 5 above)
Comment (apfitzge) [reply entry/src/entry.rs:312]: Given the diff stats it probably doesn't look like it, but I tried to keep the PR limited.
I only made a few "primary" changes:
- remove cuda API
- remove concept of "pinned" (cuda registered) vecs

Most of the changes followed from those two changes and just running checks to figure out what was now unused/broken. Some were obviously not, like user guide on using cuda, changelog, etc.


TLDR; there's more clean up to do after this PR, let's do it there.

File: perf/src/recycled_vec.rs (hunk 7 above)
Comment (apfitzge) [reply perf/src/recycled_vec.rs:24]: planned for clean up, I'm 90% sure most of this can just be removed and replaced with a `Deref` and `DerefMut` implementation on the `RecycledVec` type.

File: perf/src/recycled_vec.rs (hunk 4 above)
Comment (apfitzge) [reply perf/src/recycled_vec.rs]: 02197430403a1277397c3a7f6cd448f73bbd9160

File: perf/src/sigverify.rs (hunk 9 above)
Comment (apfitzge) [reply perf/src/sigverify.rs]: Nice catch - 17e2f1a781e77e52225f0bb0b483acb41a988094

**Comment by behzadnouri** [comment]
> recycler really only made sense in the context of pinned memory. jemalloc already keeps caches of memory that it will re-use for the packets.

There is no pinned memory here: https://github.com/anza-xyz/agave/pull/4381

> We're trying to make progress in making the chain better. 

and I am not arguing to make the chain worse. My point is:
* I am more worried about long-term sigverify scalability than an allocation or memcopy.
* If a change is potentially reducing our alternatives to address a bigger bottleneck (i.e. sigverify), lets at least do some testing first that we get anything out of it before committing to it.

**Comment by apfitzge** [comment]
> There is no pinned memory here: #4381
 
There's also no jemalloc in that benchmark though which is not indicative of a running validator's allocation performance.

That said, I'm unable to replicate the behavior on my devbox and see no difference between master (2974f02e53d3bd21e467fc902b739ad23bb605fd), reverting #4381, or adding jemalloc.
All gave me around 140k/s +/- 2k. But again not sure what options were run for that PRs testing.

> and I am not arguing to make the chain worse. My point is:
> 
>     * I am more worried about long-term sigverify scalability than an allocation or memcopy.
> 
>     * If a change is potentially reducing our alternatives to address a bigger bottleneck (i.e. sigverify), lets at least do some testing first that we get anything out of it before committing to it.

I know you're not arguing to make it worse, and did not intend to imply that - I think we both want what is best.
In my view it does not reduce our alternatives. If we need cuda we can spend the time to fix it. Have not evaluated, but jump has claimed to have significantly better cpu implementation for signature verification, which is a more immediate path to additional capacity if SV is close to becoming the major bottleneck.

File: perf/src/recycled_vec.rs (hunk 7 above)
Comment (steviez) [reply perf/src/recycled_vec.rs:24]: Sounds good / follow-on PR works for me

**Review by steviez — COMMENTED** [review]
Looks like you have another conflict that will prevent merge to master 😢

**Comment by apfitzge** [comment]
> Looks like you have another conflict that will prevent merge to master 😢

Yeah not surprising. Would like to resolve these larger conversations before fixing conflict and merging though.

**Comment by sakridge** [comment]
> edit: I apologize if my responses seem short or rude. This was all discussed on slack previously in a channel you are in.

 We did discuss this earlier in slack and it is true that no validator that we know of is using it today and it seems it will simplify reducing copies in the network pipeline and scheduler in the short term. We can still use the code in the future if we like to bring it back so I'm somewhat on board with this. I think we'll be able to create the Bytes view on top of the contiguous packet batch view if necessary. Packet batching should have benefits for CPU performance as well.

That's a bit orthogonal to the Recycler discussion, we should double check we aren't regressing anything there, maybe @lijunwangs has the benchmark setup for reproducing that.

> > What are the better solutions?
> fpga or smart nic

This is a bit hand-wavy though, what is the evidence to say it's better? Does a smartnic even exist today that can do ed25519 verify on arbitrary packet data? I had trouble finding one.

**Comment by apfitzge** [comment]
> That's a bit orthogonal to the Recycler discussion, we should double check we aren't regressing anything there, maybe @lijunwangs has the benchmark setup for reproducing that.

Sure, would appreciate any checks against regression. Any benchmark should be using jemalloc so it is similar to the validator's operation.

edit: but also to be clear, we didn't remove recycler in this PR. We only removed recyclers that were only used for the GPU code paths. 

> This is a bit hand-wavy though, what is the evidence to say it's better? Does a smartnic even exist today that can do ed25519 verify on arbitrary packet data? I had trouble finding one.

Yeah I'll admit it is and was hand wavy. I'll just use jump's numbers from this [2023 talk](https://www.youtube.com/watch?v=1oQg2_b_gv8).
- ~30ktps / cpu core
- ~1mtps / gpu (batching adds latency, 300W)
- ~1mtps / fpga (streaming, 50W)

In terms of smart nic - I'll just retract that. I don't know enough about them and was listing the alternatives discussed on slack. My understanding is that some smart nics have fpgas built in, so it may require significant reworking fd's implementation but I believe would not be too dissimilar.

**Comment by apfitzge** [comment]
> We did discuss this earlier in slack and it is true that no validator that we know of is using it today and it seems it will simplify reducing copies in the network pipeline and scheduler in the short term.

Will also expand upon this.
It gives us the possibility of moving prioritization earlier so that banking can always ingest the highest priority packets instead of needing to go in network order to find the best.

> We can still use the code in the future if we like to bring it back so I'm somewhat on board with this.

An alternative to deleting it entirely is to add a packet copy into some cuda-registered memory before we send it off for gpu verification.
That would allow us to at least isolate all the gpu stuff from the rest of the pipeline; allow CPU code to do zero-copy, and re-use the existing cuda implementation safely. We wouldn't be doing more packet copies than today, just more than CPU would after the change. This is **significantly** easier than fixing the implementation to work with non-inline memory that may or may not be contiguous.

I'm happy with either deleting or isolating; but fixing the impl to work with non-inlined data is a bigger lift.

**Comment by steviez** [comment]
> I think we'll be able to create the Bytes view on top of the contiguous packet batch view if necessary. Packet batching should have benefits for CPU performance as well.

Granted I haven't seen the branch, but this is my understanding of how this would work as well. I added a quick comment about this in https://github.com/anza-xyz/agave/pull/4803#issuecomment-2637762774, but we shouldn't do much/any worse than we currently do with `Vec<Packet>`; I guess just one extra pointer deref per packet. But, iterating packet payload will all hit that contiguous buffer.

In terms of long term scalability, another idea that has come up is huge pages. `Bytes` might get us closer to zero-copy, but huge pages would get us closer to zero-copy + zero-runtime-allocations. Given that huge pages can be done in software, this seems like something that we would want to try before telling 1400 MNB validators to figure out how to get a GPU in their rack. And, I believe huge pages are inherently pinned so whatever we cook up to work with huge pages should inherently support DMA + hardware offload.

To be clear, I'm NOT suggesting we postpone remove-CUDA + Bytes in favor of huge pages approach. Rather, I'm pointing out that there is always some better™️ optimization on the horizon, but that shouldn't stop us from pursuing short/medium term improvements. So, I'm in favor of the (possibly temporary) removal of CUDA support that this PR makes

**Comment by sakridge** [comment]
> Yeah I'll admit it is and was hand wavy. I'll just use jump's numbers from this [2023 talk](https://www.youtube.com/watch?v=1oQg2_b_gv8).
> 
> * ~30ktps / cpu core
> * ~1mtps / gpu (batching adds latency, 300W)
> * ~1mtps / fpga (streaming, 50W)

I've benchmarked Nvidia 3090 (released 2020) at around 3m/s and 4090 (released in 2022) at 10m/s, so I'm not sure how much I trust these numbers for GPU or what implementation they are using. There's no type of GPU, code or concrete benchmarks presented. I agree GPU likely has more latency, from my testing it does seem to be in the 5-10ms range for batch sizes which exceed the CPU speed which is not great, but is still workable within our constraints. FPGAs likely have a power advantage which not clear how much that really matters, but I think there probably needs to be a more rigorous analysis comparing like-for-like since there are many models of FPGAs and GPUs. I don't think one can take some spitballed numbers from a slide to do a good analysis and the situation changes since new GPUs and FPGAs are released all the time with different software stacks and whatnot that can improve latency and overheads. There are costs and availability concerns especially with FGPAs where the expensive ones can cost $20k+ each and aren't common in datacenters. Anyway, I think my position is I don't really know which hardware wins here and it's nice to have options. I also think it could be somewhat likely that FPGAs have similar memory constraints to the GPU in that to get the best performance you would setup DMA copy engines which can't really deal with CPU page-faults well or at all without a huge complexity hit or a large list of memory ranges that you would need for a highly fragmented copy to device memory.

I think the extra copy for me would be fine to introduce for the GPU for now and keep the path in. I think it will be somewhat harder to add it back in later if we completely remove it.

**Comment by behzadnouri** [comment]
> create the Bytes view on top of the contiguous packet batch

I think the starting presumption here is that moving to `Bytes` is a good thing and improves performance.
But I am not even confident that is true:

* `Bytes` does dynamic dispatch which is pretty slow, particularly so in certain runtime access patterns.
* We already use `Recycler` for packets which `Bytes` is not compatible with (unless we do memcopies anyways). So not even sure we will do fewer allocations or memcopies with `Bytes`.
* `Bytes` is pushing out the gpu code apparently. Again, sigverify is a bigger bottleneck than an allocation or memcopy.
* `Bytes` does not work with `[u8; N]`, so you are always forced an extra redirection (`Packet` is just a simple `[u8; N]` wrapper). `Bytes` also does not work with `Arc<Vec<u8>>` either.

Moving to `Bytes` needs pretty big and widespread changes (including this one), and if they are committed to master, in 2 or 3 months it would be practically impossible to revert them (due to merge conflicts and code diverging off).

So why not develop these changes first on an off-master branch, and get some reliable estimated numbers that performance improvements from `Bytes` (if any) does indeed justify the downsides?

If anything, an off-master branch would allow to iterate much faster.

**Comment by alexpyattaev** [comment]
Adding some insight form someone with several years of FPGA code development:

- Maintaining FPGA code is pretty nightmarish, most tools are closed-source and terrible. Build times are a horror story like no other. CI tooling is also largely proprietary and terrible.
- Debugging FPGA/CPU binding/driver code is unfun. Very very unfun. Let us not do it.
- Pretty much all good FPGA code is platform-specific. What works well on Altera/Intel may not work so well on Xilinx. Port mapping (i.e. interfacing with outside world) is always platform specific.
- Talking to the FPGA accelerator generally requires kernel driver of some sort (as you need to map some address space to be accessible by the device via PCIe) and/or a proprietary SDK (which is terrible). All of that is nasty and wildly unstable. 
- Contrary to what was mentioned above, FPGA accelerators do not necessarily require any particularly fancy memory layout of the input data, as long as there is sufficient bus bandwidth to copy things from main memory to the FPGA. FPGA-enabled smart NICs work on a per-packet basis and as such do not care how we sotre packets in agave.

Based on the above, my recommendation is to stay away from FPGA code if at all possible. Given that GPU acceleration can provide far more throughput than CPU with reasonable latencies, my suggestion would be to rely on those rather than FPGAs. 

For me it seems that sigverify as such does not really require much context to work (and whatever context is required can be easily provided over RPC). So doing sigverify in a separate process (or even on a separate host) is not so hard. Having a dedicated process that uses e.g. a GPU to bulk sigverify all passing packets and just drop all the invalid ones makes far more sense to me than having to talk to the GPU from within agave:
 - we would get all the same perf benefits but without the added complexity in the agave codebase
 - if the GPU accelerator process segfaults due to some silly driver bug, it can be restarted in seconds rather than 15 minutes
 - one can have several of those GPU boxes with failover set up
 - for the operators it would also add the flexibility of being able to use a normal "gaming PCs" as the GPU accelerator frontends for the validator rather than fancy expensive GPU nodes tuned for ML applications
